// @udecode/utils/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  IS_APPLE: () => IS_APPLE,
  escapeRegExp: () => escapeRegExp,
  findHtmlParentElement: () => findHtmlParentElement,
  getHandler: () => getHandler,
  hexToBase64: () => hexToBase64,
  isDefined: () => isDefined,
  isNull: () => isNull,
  isUndefined: () => isUndefined,
  isUndefinedOrNull: () => isUndefinedOrNull,
  isUrl: () => isUrl,
  mergeProps: () => mergeProps,
  sanitizeUrl: () => sanitizeUrl
});
module.exports = __toCommonJS(src_exports);

// src/environment.ts
var IS_APPLE = typeof navigator !== "undefined" && /Mac OS X/.test(navigator.userAgent);

// src/escapeRegexp.ts
var escapeRegExp = (text) => {
  return text.replaceAll(/[#$()*+,.?[\\\]^s{|}-]/g, "\\$&");
};

// src/findHtmlParentElement.ts
var findHtmlParentElement = (el, nodeName) => {
  if (!el || el.nodeName === nodeName) {
    return el;
  }
  return findHtmlParentElement(el.parentElement, nodeName);
};

// src/getHandler.ts
var getHandler = (cb, ...args) => () => {
  cb == null ? void 0 : cb(...args);
};

// src/hexToBase64.ts
var hexToBase64 = (hex) => {
  const hexPairs = hex.match(/\w{2}/g) || [];
  const binary = hexPairs.map(
    (hexPair) => String.fromCodePoint(Number.parseInt(hexPair, 16))
  );
  return btoa(binary.join(""));
};

// src/isUrl.ts
var protocolAndDomainRE = /^(?:\w+:)?\/\/(\S+)$/;
var emailLintRE = /mailto:([^?\\]+)/;
var localhostDomainRE = /^localhost[\d:?]*(?:[^\d:?]\S*)?$/;
var nonLocalhostDomainRE = /^[^\s.]+\.\S{2,}$/;
var isUrl = (string) => {
  if (typeof string !== "string") {
    return false;
  }
  const generalMatch = string.match(protocolAndDomainRE);
  const emailLinkMatch = string.match(emailLintRE);
  const match = generalMatch || emailLinkMatch;
  if (!match) {
    return false;
  }
  const everythingAfterProtocol = match[1];
  if (!everythingAfterProtocol) {
    return false;
  }
  try {
    new URL(string);
  } catch (e) {
    return false;
  }
  return localhostDomainRE.test(everythingAfterProtocol) || nonLocalhostDomainRE.test(everythingAfterProtocol);
};

// src/mergeProps.ts
var mergeProps = (props, overrideProps, {
  handlerKeys,
  handlerQuery = (key) => key.indexOf("on") === 0
} = {}) => {
  const map = /* @__PURE__ */ new Map();
  const acc = {};
  const mapProps = (_props) => {
    if (!_props)
      return;
    Object.entries(_props).forEach(([key, value]) => {
      var _a;
      if ((!handlerKeys || handlerKeys.includes(key)) && (!handlerQuery || handlerQuery(key)) && typeof value === "function") {
        if (!map.has(key)) {
          map.set(key, []);
        }
        (_a = map.get(key)) == null ? void 0 : _a.push(value);
        acc[key] = (...args) => {
          var _a2;
          (_a2 = map.get(key)) == null ? void 0 : _a2.forEach((fn) => fn(...args));
        };
      } else {
        acc[key] = value;
      }
    });
  };
  mapProps(props);
  mapProps(overrideProps);
  return acc;
};

// src/sanitizeUrl.ts
var sanitizeUrl = (url, { allowedSchemes, permitInvalid = false }) => {
  if (!url)
    return null;
  let parsedUrl = null;
  try {
    parsedUrl = new URL(url);
  } catch (e) {
    return permitInvalid ? url : null;
  }
  if (allowedSchemes && !allowedSchemes.includes(parsedUrl.protocol.slice(0, -1))) {
    return null;
  }
  return parsedUrl.href;
};

// src/type-utils.ts
var isUndefined = (obj) => obj === void 0;
var isNull = (obj) => obj === null;
var isUndefinedOrNull = (obj) => isUndefined(obj) || isNull(obj);
var isDefined = (arg) => !isUndefinedOrNull(arg);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  IS_APPLE,
  escapeRegExp,
  findHtmlParentElement,
  getHandler,
  hexToBase64,
  isDefined,
  isNull,
  isUndefined,
  isUndefinedOrNull,
  isUrl,
  mergeProps,
  sanitizeUrl
});
